sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -1162,41 +1140,27 @@ case class Unhex(child: Expression, failOnError: Boolean = false)
   override def dataType: DataType = BinaryType
 
   protected override def nullSafeEval(num: Any): Any = {
-    val result = Hex.unhex(num.asInstanceOf[UTF8String].getBytes)
-    if (failOnError && result == null) {
-      // The failOnError is set only from `ToBinary` function - hence we might safely set `hint`
-      // parameter to `try_to_binary`.
-      throw QueryExecutionErrors.invalidInputInConversionError(
-        BinaryType,
-        num.asInstanceOf[UTF8String],
-        UTF8String.fromString("HEX"),
-        "try_to_binary")
+    try {
+      Hex.unhex(num.asInstanceOf[UTF8String].toString)
```

Comment:
The improvement to the `unhex` function seems effective in isolated cases, but IIRC, `UTF8String.toString` always incurs an additional Array Copy overhead compared to `UTF8String.getBytes`. Is this really better? 

Therefore, should we add a offline test to evaluate the effectiveness of the improvement for the `Unhex` sql function, possibly focusing on both short and long strings?

Reply:
Good point, let me implement unhex directly on bytes